modules/core/src/test/java/org/apache/ignite/internal/util/DistributedProcessCoordinatorLeftTest.java

```diff
@@ -0,0 +1,158 @@
+
+/**
+ * Tests {@link DistributedProcess} in case of coordinator node left.
+ */
+public class DistributedProcessCoordinatorLeftTest extends GridCommonAbstractTest {
+    /** */
+    public static final long TIMEOUT = 20_000L;
+
+    /** */
+    public static final int NODES_CNT = 3;
+
+    /** Latch to send single message on node left. */
+    private final CountDownLatch nodeLeftLatch = new CountDownLatch(NODES_CNT - 1);
+
+    /** Latch to await sending single messages to a failed coordinator. */
+    private final CountDownLatch msgSendLatch = new CountDownLatch(NODES_CNT - 1);
+
+    /** Failure handler invocation flag. */
+    private final AtomicBoolean failure = new AtomicBoolean();
+
+    /** */
+    private final ListeningTestLogger listeningLog = new ListeningTestLogger(true, log);
+
+    /** {@inheritDoc} */
+    @Override protected IgniteConfiguration getConfiguration(String igniteInstanceName) throws Exception {
+        IgniteConfiguration cfg = super.getConfiguration(igniteInstanceName);
+
+        cfg.setGridLogger(listeningLog);
+
+        cfg.setLocalEventListeners(Collections.singletonMap(event -> {
+            nodeLeftLatch.countDown();
+
+            try {
+                msgSendLatch.await();
+            }
+            catch (InterruptedException e) {
+                fail("Unexpected interrupt.");
+            }
+
+            return false;
+        }, new int[] {EVT_NODE_LEFT, EVT_NODE_FAILED}));
+
+        cfg.setFailureHandler(new FailureHandler() {
+            @Override public boolean onFailure(Ignite ignite, FailureContext failureCtx) {
+                failure.set(true);
+
+                return false;
+            }
+        });
+
+        return cfg;
+    }
+
+    /**
+     * Tests that coordinator failing during sending single result not cause node failure and the process finishes.
+     *
+     * <ol>
+     *  <li>Start new process of {@link DistributedProcess}.</li>
+     *  <li>The coordinator fails.</li>
+     *  <li>Nodes try to send a single message to the not-alive coordinator.</li>
+     *  <li>{@link DistributedProcess} process a node left event and reinitialize a new coordinator.</li>
+     *  <li>Process finishes.</li>
+     * </ol>
+     *
+     * @throws Exception If failed.
+     */
+    @Test
+    public void testCoordinatorFailed() throws Exception {
+        startGrids(NODES_CNT);
+
+        CountDownLatch startLatch = new CountDownLatch(NODES_CNT);
+        CountDownLatch finishLatch = new CountDownLatch(NODES_CNT - 1);
+
+        HashMap<String, DistributedProcess<Integer, Integer>> processes = new HashMap<>();
+
+        int processRes = 1;
+
+        for (Ignite grid : G.allGrids()) {
+            DistributedProcess<Integer, Integer> dp = new DistributedProcess<>(((IgniteEx)grid).context(), TEST_PROCESS,
+                req -> GridTestUtils.runAsync(() -> {
+                    startLatch.countDown();
```

Comment:
Should we release latch only after the `single-message` future completed? It's still possible that a new coordinator will be assigned prior to the single message sent.

Reply:
The mentioned `startLatch` is to make sure that process started and we can drop coordinator.
A new coordinator can't be reassigned because the event queue is blocked by `msgSendLatch` until single messages sent.

Reply:
The single message future must be completed prior to the event queue processing released. Is it true for now? Currently, this will happen at the same time and we will have a race with a new coordinator assigned to the `Process`.

Reply:
Yes, it's true. 'msgSendLatch' guarantees this. The coordinator will be changed at the event thread after releasing the latch.